File: Backend/load_test.py
#!/usr/bin/env python3
"""
⚡ Load Testing Script for Lugn & Trygg Backend
Tests API performance under 1000 concurrent users

Requirements:
    pip install locust requests

Usage:
    # Start load test web UI
    locust -f load_test.py --host=http://localhost:5001
    
    # Then open http://localhost:8089
    # Set: 1000 users, spawn rate 100/sec
    
    # Or run headless:
    locust -f load_test.py --host=http://localhost:5001 \\
           --users 1000 --spawn-rate 100 --run-time 5m --headless
"""

from locust import HttpUser, task, between
import random
import json
from datetime import datetime

class LugnTryggUser(HttpUser):
    """
    Simulates a real user of Lugn & Trygg
    """
    wait_time = between(1, 5)  # Wait 1-5 seconds between requests
    
    def on_start(self):
        """Called when a user starts - simulate login"""
        self.user_id = f"test_user_{random.randint(1, 10000)}"
        self.token = None
        
        # Try to login (mock for now)
        # In real test, you'd call actual auth endpoint
        self.token = "mock_jwt_token"
    
    @task(10)
    def health_check(self):
        """
        Test health endpoint - most frequent
        Weight: 10 (10% of requests)
        """
        self.client.get("/api/health")
    
    @task(8)
    def get_moods(self):
        """
        Fetch mood logs - common operation
        Weight: 8 (8% of requests)
        """
        headers = {"Authorization": f"Bearer {self.token}"} if self.token else {}
        self.client.post(
            "/api/mood/get",
            json={"user_id": self.user_id},
            headers=headers
        )
    
    @task(5)
    def log_mood(self):
        """
        Log a new mood - frequent user action
        Weight: 5 (5% of requests)
        """
        headers = {"Authorization": f"Bearer {self.token}"} if self.token else {}
        mood_data = {
            "user_id": self.user_id,
            "mood": random.choice(["happy", "sad", "anxious", "calm", "energetic"]),
            "score": random.randint(1, 10),
            "note": "Load test mood log",
            "timestamp": datetime.utcnow().isoformat()
        }
        self.client.post(
            "/api/mood/log",
            json=mood_data,
            headers=headers
        )
    
    @task(5)
    def get_weekly_analysis(self):
        """
        Get weekly mood analysis
        Weight: 5 (5% of requests)
        """
        headers = {"Authorization": f"Bearer {self.token}"} if self.token else {}
        self.client.post(
            "/api/mood/weekly-analysis",
            json={"user_id": self.user_id},
            headers=headers
        )
    
    @task(3)
    def ai_chat(self):
        """
        Chat with AI - moderate frequency
        Weight: 3 (3% of requests)
        """
        headers = {"Authorization": f"Bearer {self.token}"} if self.token else {}
        self.client.post(
            "/api/ai/chat",
            json={
                "user_id": self.user_id,
                "message": "How can I improve my mood today?",
                "session_id": f"session_{self.user_id}"
            },
            headers=headers
        )
    
    @task(3)
    def get_chat_history(self):
        """
        Fetch chat history
        Weight: 3 (3% of requests)
        """
        headers = {"Authorization": f"Bearer {self.token}"} if self.token else {}
        self.client.post(
            "/api/ai/history",
            json={"user_id": self.user_id},
            headers=headers
        )
    
    @task(2)
    def get_memories(self):
        """
        Fetch memory list
        Weight: 2 (2% of requests)
        """
        headers = {"Authorization": f"Bearer {self.token}"} if self.token else {}
        self.client.post(
            "/api/memory/list",
            json={"user_id": self.user_id},
            headers=headers
        )
    
    @task(2)
    def get_leaderboard(self):
        """
        Get referral leaderboard
        Weight: 2 (2% of requests)
        """
        self.client.get("/api/referral/leaderboard?limit=100")
    
    @task(1)
    def predictive_analytics(self):
        """
        Get mood predictions - CPU intensive
        Weight: 1 (1% of requests)
        """
        headers = {"Authorization": f"Bearer {self.token}"} if self.token else {}
        self.client.post(
            "/api/predictive/mood-forecast",
            json={"user_id": self.user_id, "days": 7},
            headers=headers
        )
    
    @task(1)
    def metrics_check(self):
        """
        Check system metrics
        Weight: 1 (1% of requests)
        """
        self.client.get("/api/metrics")

# Performance targets for 1000 users
"""
Target Metrics:
- Response Time (95th percentile): < 500ms
- Error Rate: < 1%
- Requests/sec: > 200
- Failed Requests: < 10

Success Criteria:
✅ Health endpoint: < 50ms
✅ Mood log: < 200ms
✅ AI chat: < 2000ms (AI is slow)
✅ Database queries: < 100ms
"""

File: Backend/load-test.js
/**
 * k6 Load Test for Lugn & Trygg Backend
 * 
 * Installation: 
 *   Windows: choco install k6
 *   Mac: brew install k6
 *   Linux: sudo apt install k6
 * 
 * Run:
 *   k6 run load-test.js
 *   k6 run load-test.js --vus 100 --duration 1m  (100 concurrent users, 1 minute)
 *   k6 run load-test.js --vus 1000 --duration 5m (1000 concurrent users, 5 minutes)
 */

import http from 'k6/http';
import { check, sleep } from 'k6';
import { Rate, Trend } from 'k6/metrics';

// Custom metrics
const errorRate = new Rate('errors');
const loginDuration = new Trend('login_duration');
const moodLogDuration = new Trend('mood_log_duration');

// Test configuration
export const options = {
    stages: [
        { duration: '30s', target: 50 },   // Ramp up to 50 users
        { duration: '1m', target: 100 },   // Stay at 100 users
        { duration: '2m', target: 500 },   // Ramp up to 500 users
        { duration: '2m', target: 1000 },  // Ramp up to 1000 users (target)
        { duration: '1m', target: 0 },     // Ramp down
    ],
    thresholds: {
        'http_req_duration': ['p(95)<2000'],  // 95% of requests under 2s
        'http_req_failed': ['rate<0.01'],      // Less than 1% failures
        'errors': ['rate<0.05'],               // Less than 5% errors
    },
};

const BASE_URL = __ENV.BASE_URL || 'http://localhost:5001';

// Test data
const testUsers = [
    { email: 'loadtest1@test.com', password: 'TestPassword123!' },
    { email: 'loadtest2@test.com', password: 'TestPassword123!' },
    { email: 'loadtest3@test.com', password: 'TestPassword123!' },
];

export function setup() {
    // Health check before starting
    const res = http.get(`${BASE_URL}/api/health`);
    check(res, {
        'Health check passed': (r) => r.status === 200,
    });
    
    console.log(`Starting load test against ${BASE_URL}`);
    return { startTime: new Date().toISOString() };
}

export default function () {
    const user = testUsers[Math.floor(Math.random() * testUsers.length)];
    
    // 1. Health check (lightweight)
    const healthRes = http.get(`${BASE_URL}/api/health`);
    check(healthRes, { 'Health OK': (r) => r.status === 200 });
    
    // 2. Login attempt
    const loginStart = new Date();
    const loginRes = http.post(`${BASE_URL}/api/auth/login`, JSON.stringify({
        email: user.email,
        password: user.password,
    }), {
        headers: { 'Content-Type': 'application/json' },
    });
    loginDuration.add(new Date() - loginStart);
    
    const loginSuccess = check(loginRes, {
        'Login status 200 or 401': (r) => r.status === 200 || r.status === 401,
    });
    
    if (!loginSuccess) {
        errorRate.add(1);
        console.error(`Login failed: ${loginRes.status} - ${loginRes.body}`);
        return;
    }
    
    // If login succeeded, test authenticated endpoints
    if (loginRes.status === 200) {
        const body = JSON.parse(loginRes.body);
        const token = body.access_token;
        
        const authHeaders = {
            'Content-Type': 'application/json',
            'Authorization': `Bearer ${token}`,
        };
        
        // 3. Log mood
        const moodStart = new Date();
        const moodRes = http.post(`${BASE_URL}/api/mood/log`, JSON.stringify({
            value: Math.floor(Math.random() * 10) + 1,
            note: 'Load test mood entry',
        }), { headers: authHeaders });
        moodLogDuration.add(new Date() - moodStart);
        
        check(moodRes, {
            'Mood log succeeded': (r) => r.status === 200 || r.status === 201,
        });
        
        // 4. Get mood history
        const historyRes = http.get(`${BASE_URL}/api/mood/history`, { headers: authHeaders });
        check(historyRes, {
            'Mood history retrieved': (r) => r.status === 200,
        });
        
        // 5. Dashboard stats
        const dashRes = http.get(`${BASE_URL}/api/dashboard/stats`, { headers: authHeaders });
        check(dashRes, {
            'Dashboard stats retrieved': (r) => r.status === 200,
        });
    }
    
    sleep(1); // Wait 1 second between iterations
}

export function teardown(data) {
    console.log(`Load test completed. Started at: ${data.startTime}`);
}
